e2e/dractions/actionsdiscoveredapps.go

```diff
@@ -0,0 +1,170 @@
+		return err
+	}
+
+	util.Ctx.Log.Info("create drpc " + drpcName)
+
+	clusterName := drpolicy.Spec.DRClusters[0]
+
+	drpc := generateDRPCDiscoveredApps(name, RamenOpsNs, clusterName, drPolicyName, placementName, appname, namespace)
+	if err = createDRPC(util.Ctx.Hub.CtrlClient, drpc); err != nil {
+		return err
+	}
+
+	// wait for drpc ready
+	return waitDRPCReady(util.Ctx.Hub.CtrlClient, RamenOpsNs, drpcName)
+}
+
+// remove DRPC
+// update placement annotation
+func DisableProtectionDiscoveredApps(w workloads.Workload, d deployers.Deployer) error {
+	util.Ctx.Log.Info("enter DisableProtectionDiscoveredApps DisableProtection")
+
+	name := GetCombinedName(d, w)
+
+	placementName := name
+	drpcName := name
+
+	client := util.Ctx.Hub.CtrlClient
+
+	util.Ctx.Log.Info("delete drpc " + drpcName)
+
+	if err := deleteDRPC(client, RamenOpsNs, drpcName); err != nil {
+		return err
+	}
+
+	if err := waitDRPCDeleted(client, RamenOpsNs, drpcName); err != nil {
+		return err
+	}
+
+	// delete placement
+	if err := deployers.DeletePlacement(placementName, RamenOpsNs); err != nil {
+		return err
+	}
+
+	return deployers.DeleteManagedClusterSetBinding(deployers.McsbName, RamenOpsNs)
+}
+
+func FailoverDiscoveredApps(w workloads.Workload, d deployers.Deployer) error {
+	util.Ctx.Log.Info("enter DRActions FailoverDiscoveredApps")
+
+	return failoverRelocateAction(w, d, ramen.ActionFailover)
+}
+
+func RelocateDiscoveredApps(w workloads.Workload, d deployers.Deployer) error {
+	util.Ctx.Log.Info("enter DRActions RelocateDiscoveredApps")
+
+	return failoverRelocateAction(w, d, ramen.ActionRelocate)
+}
+
+// nolint:funlen,cyclop
+func failoverRelocateAction(w workloads.Workload, d deployers.Deployer, action ramen.DRAction) error {
+	name := GetCombinedName(d, w)
+	namespace := name
+
+	drPolicyName := DefaultDRPolicyName
+	drpcName := name
+	hubClient := util.Ctx.Hub.CtrlClient
+
+	// here we expect drpc should be ready before action
+	if err := waitDRPCReady(hubClient, RamenOpsNs, drpcName); err != nil {
+		return err
+	}
+
+	util.Ctx.Log.Info("get drpc " + drpcName)
+
+	drpc, err := getDRPC(hubClient, RamenOpsNs, drpcName)
+	if err != nil {
+		return err
+	}
+
+	util.Ctx.Log.Info("get drpolicy " + drPolicyName)
+
+	drpolicy, err := getDRPolicy(hubClient, drPolicyName)
+	if err != nil {
+		return err
+	}
+
+	currentCluster, err := getCurrentCluster(hubClient, RamenOpsNs, name)
+	if err != nil {
+		return err
+	}
+
+	drClient := getDRClusterClient(currentCluster, drpolicy)
+
+	targetCluster, err := getTargetCluster(hubClient, RamenOpsNs, name, drpolicy)
+	if err != nil {
+		return err
+	}
+
+	util.Ctx.Log.Info(strings.ToLower(string(action)) + " to cluster: " + targetCluster)
+
+	drpc.Spec.Action = action
+	if action == ramen.ActionFailover {
+		drpc.Spec.FailoverCluster = targetCluster
+	} else {
+		drpc.Spec.PreferredCluster = targetCluster
+	}
+
+	util.Ctx.Log.Info("update drpc " + drpcName)
+
+	if err = updateDRPC(hubClient, drpc); err != nil {
+		return err
+	}
```

Comment:
Until this point the action is the same for discovered and managed apps, we should refactor this out.

Reply:
refactored according to the suggestion